File: 2022/day1/c++/part2.cpp
#include <fstream>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

using namespace std;

int main() {
  ifstream input("../input.txt");

  if (!input.is_open()) {
    cerr << "Unable to open input file\n";

    return 1;
  }

  vector<size_t> all_calories;

  string line;
  size_t calories = 0;
  while (getline(input, line)) {
    if (line == "") {
      all_calories.push_back(calories);

      calories = 0;
      continue;
    }

    calories += stoi(line);
  }

  sort(all_calories.begin(), all_calories.end(), greater<size_t>());
  all_calories.resize(3);

  size_t most_calories_top3 = reduce(all_calories.begin(), all_calories.end());

  cout << most_calories_top3 << '\n';
}

File: 2022/day1/javascript/part2.mjs
import { readFileSync } from 'node:fs';

const input = readFileSync('../input.txt', 'utf-8');

const calorieGroups = input.split(/(?:\r?\n){2}/);

const mostCaloriesTop3 = calorieGroups
  .map((calories) =>
    calories
      .split(/\r?\n/)
      .reduce((total, calorie) => total + parseInt(calorie), 0)
  )
  .sort((a, b) => b - a)
  .slice(0, 3)
  .reduce((all, calorie) => all + calorie, 0);

console.log(mostCaloriesTop3);
